app/models/framework_response.rb

```diff
@@ -45,4 +50,52 @@ def build_flags
       end
     end
   end
+
+  def clear_dependent_values_and_flags!(old_value)
+    return unless old_value != value
+
+    dependent_ids = dependents.includes(:framework_question).reject { |dependent| option_selected?(dependent.framework_question.dependent_value) }
+
+    update_dependent_responses!(dependent_ids)
+  end
+
+  def update_dependent_responses!(dependent_ids)
+    return unless dependent_ids.any?
+
+    descendants = descendants_tree(dependent_ids).includes(:framework_question, :flags).map do |descendant|
+      descendant.assign_attributes(
+        value: nil,
+        responded: false,
+        flags: [],
+      )
+
+      descendant
+    end
+
+    FrameworkResponse.import(descendants, validate: false, recursive: true, all_or_none: true, on_duplicate_key_update: %i[value_json value_text responded])
+  end
+
+  def descendants_tree(ids)
+    FrameworkResponse
+      .where("framework_responses.id IN (#{recursive_tree})", ids)
+  end
+
+  def recursive_tree
+    # build full descendants tree
+    <<-SQL
```

Comment:
This approach is pretty low-level sql and will be hard to maintain... is than an alternative more active-record friendly approach? could the acts_as_tree gem possibly help?

Reply:
I usually try to avoid building complex low level queries and default to using active record DSL provided, however the ability to build a recursive function wasn't possible unfortunately. I had a look at a few gems that could help, however `acts_as_tree` gem doe not support this kind of querying and only normally walks the tree. Another gem that does support this seems to not be actively maintained https://github.com/chrisroberts/acts_as_sane_tree. I did look into the ancestry gem initially but it did not play well with bulk creation of dependencies so I had to abandon it.

I kept this query quite simple in structure, where the only thing that is different is that it utilises the [recursive functionality](https://en.wikipedia.org/wiki/Hierarchical_and_recursive_queries_in_SQL#Common_table_expression) in [Postgres](https://www.postgresql.org/docs/12/queries-with.html).  It starts by selecting the root nodes required, and then the next section defines the recursive query to retrieve the next nodes. I think as long as we keep this query documented well and relatively simple in structure it can be the tradeoff.

Reply:
Thanks for looking into alternatives 👍